src/backends/r1cs/mod.rs

```diff
@@ -0,0 +1,557 @@
+    pub fn private_input_number(&self) -> usize {
+        self.witness_vars.len() - self.public_inputs.len() - self.public_outputs.len()
+    }
+
+    // Helper function to convert &[u64] to BigInt
+    fn bigint_from_u64_slice(slice: &[u64]) -> BigInt {
+        let mut bytes = Vec::new();
+        for &num in slice.iter().rev() { // Reverse the slice to match the big-endian byte order
+            bytes.extend(&num.to_be_bytes());
+        }
+        BigInt::from_bytes_be(Sign::Plus, &bytes)
+    }
+}
+
+#[derive(Debug)]
+/// An intermediate struct for SnarkjsExporter to reorder the witness and convert to snarkjs format 
+pub struct GeneratedWitness {
+    pub witness: HashMap<usize, Fr>,
+}
+
+impl Backend for R1csBls12_381 {
+    type Field = Fr;
+
+    type GeneratedWitness = GeneratedWitness;
+
+    fn witness_vars(&self) -> &std::collections::HashMap<usize, crate::var::Value<Self>> {
+        &self.witness_vars
+    }
+
+    fn poseidon() -> crate::imports::FnHandle<Self> {
+        poseidon
+    }
+
+    fn new_internal_var(
+        &mut self,
+        val: crate::var::Value<Self>,
+        span: crate::constants::Span,
+    ) -> CellVar {
+        let var = CellVar::new(self.next_variable, span);
+        self.next_variable += 1;
+
+        // store it in the circuit_writer
+        self.witness_vars.insert(var.index, val);
+
+        var
+    }
+
+    fn add_constant(
+        &mut self,
+        //todo: do we need this?
+        label: Option<&'static str>,
+        value: Self::Field,
+        span: crate::constants::Span,
+    ) -> CellVar {
+        let x = self.new_internal_var(Value::Constant(value), span);
+        self.assert_eq_const(&x, value, span);
+
+        x
+    }
+
+    /// Final checks for generating the circuit
+    /// todo: we might need to rethink about this interface
+    /// - private_input_indices are not needed in this r1cs backend
+    /// - main_span could be better initialized with the backend, so it doesn't have to pass in here?
+    /// - we might just need the returned_cells argument, as the backend can record the public outputs itself?
+    fn finalize_circuit(
+        &mut self,
+        public_output: Option<crate::var::Var<Self::Field>>,
+        returned_cells: Option<Vec<CellVar>>,
+        _private_input_indices: Vec<usize>,
+        _main_span: crate::constants::Span,
+    ) -> crate::error::Result<()> {
+        // store the return value in the public input that was created for that
+        if let Some(public_output) = public_output {
+            let cvars = &public_output.cvars;
+
+            for (pub_var, ret_var) in cvars.clone().iter().zip(returned_cells.unwrap()) {
+                // replace the computation of the public output vars with the actual variables being returned here
+                let var_idx = pub_var.idx().unwrap();
+                let prev = self
+                    .witness_vars
+                    .insert(var_idx, Value::PublicOutput(Some(ret_var)));
```

Comment:
the way I would do it is just do `self.witness_vars[var_idx] = Value::PublicOutput(Some(ret_var));